File: ArraysAndHashing/1_ContainsDuplicate.cs
/**
https://leetcode.com/problems/contains-duplicate/

Given an integer array nums, return true if any value appears at least twice in the array, and return false if every element is distinct.
*/

public bool ContainsDuplciate(int[] nums) {
    HashSet<int> track = new HashSet<int>();
    foreach (int n in nums) {
        if (track.Contains(n)) return true;
        track.Add(n);
    }
    return false;
}

/**
Algorithm:
Use a hashset to keep track of the numbers already visited, for each number in the nums array, check if the hashset already contains that number.

Space: O(n), all elements stored in a hashset
Time: O(n), full nums array is iterated through
*/

File: ArraysAndHashing/Java/1_ContainsDuplicate.java
/**
https://leetcode.com/problems/contains-duplicate/

Given an integer array nums, return true if any value appears at least twice in the array, and return false if every element is distinct.
*/

class Solution {
    public boolean containsDuplicate(int[] nums) {
        HashSet<Integer> track = new HashSet();
        for (int n : nums) {
            if (track.contains(n)) return true;
            track.add(n);
        }
        return false;
    }
}

/**
Algorithm: store the visited numbers in a hashset, return false if the set already contains that number

Space: O(n), all elements stored in the hashset
Time: O(n), all elements in the num array iterated through
*/